TP Broken in version 0.0.9+

TP stopped working in version 0.0.9 and newer. When rolling back to 0.0.8, it works normally. 

Model: "EleutherAI/gpt-neo-1.3B" with tp_degree=8


```
File "/usr/local/lib/python3.10/site-packages/neuronx_distributed/parallel_layers/layers.py", line 180, in __init__
    ) = EmbeddingUtility.range_from_global_vocab_size(
  File "/usr/local/lib/python3.10/site-packages/neuronx_distributed/parallel_layers/utils.py", line 62, in range_from_global_vocab_size
    per_partition_vocab_size = divide(global_vocab_size, world_size)
  File "/usr/local/lib/python3.10/site-packages/neuronx_distributed/parallel_layers/utils.py", line 17, in divide
    ensure_divisibility(numerator, denominator)
  File "/usr/local/lib/python3.10/site-packages/neuronx_distributed/parallel_layers/utils.py", line 9, in ensure_divisibility
    assert numerator % denominator == 0, "
{}
 is not divisible by
{}
".format(
AssertionError: 50257 is not divisible by 8
```

--- a/optimum/neuron/accelerate/accelerator.py
+++ b/optimum/neuron/accelerate/accelerator.py
@@ -109,7 +109,7 @@ class NeuronAccelerator(Accelerator):
                 tp_size = 1
             else:
                 tp_size = int(use_neuronx_distributed_tp)
-            tp_plugin = TensorParallelismPlugin(tensor_parallel_size=tp_size)
+            tp_plugin = TensorParallelismPlugin(tensor_parallel_size=tp_size, parallelize_embeddings=True)
         self._model_cpu_parameters_to_xla = {}
 
         if tp_plugin.should_parallelize:

--- a/optimum/neuron/accelerate/accelerator.py
+++ b/optimum/neuron/accelerate/accelerator.py
@@ -30,6 +30,8 @@ from accelerate.utils import DistributedType
 from torch.utils.data import DataLoader
 from torch.utils.data.distributed import DistributedSampler
 
+from optimum.neuron.utils.patching import ModelPatcher
+
 from ...utils import logging
 from ..distributed import Parallelizer, ParallelizersManager
 from ..distributed.utils import ZeroRedundancyOptimizerCompatibleWithTensorParallelism
@@ -305,8 +307,17 @@ class NeuronAccelerator(Accelerator):
                 model.to(torch.bfloat16)
             else:
                 model.to(torch.float32)
-            parallel_layers.move_model_to_device(model, self.device)
-            model.tie_weights()
+
+            def _tie_or_clone_weights_for_tp(self, output_embeddings, input_embeddings):
+                """Tie or clone module weights depending of whether we are using TorchScript or not"""
+                output_embeddings.weight = input_embeddings.weight
+                if hasattr(output_embeddings, "out_features") and hasattr(input_embeddings, "num_embeddings"):
+                    output_embeddings.out_features = input_embeddings.num_embeddings
+
+            with ModelPatcher(patching_specs=[(model, "_tie_or_clone_weights", _tie_or_clone_weights_for_tp)]):
+                model.tie_weights()
+                parallel_layers.move_model_to_device(model, self.device)
+                model.tie_weights()
             self._model_cpu_parameters_to_xla[id(model)] = dict(zip(cpu_ids, model.parameters()))
             device_placement = False
         return super().prepare_model(model, device_placement=device_placement, evaluation_mode=evaluation_mode)

--- a/optimum/neuron/accelerate/utils/dataclasses.py
+++ b/optimum/neuron/accelerate/utils/dataclasses.py
@@ -141,6 +141,7 @@ class NeuronFullyShardedDataParallelPlugin(FullyShardedDataParallelPlugin):
 @dataclass
 class TensorParallelismPlugin:
     tensor_parallel_size: int = 1
+    parallelize_embeddings: bool = True
 
     def __post_init__(self):
         if self.tensor_parallel_size < 1:
@@ -162,6 +163,7 @@ class TensorParallelismPlugin:
             model,
             orig_to_parallel=orig_to_parallel,
             device=device,
+            parallelize_embeddings=self.parallelize_embeddings,
         )
         if return_orig_to_parallel:
             return parallelized_model, orig_to_parallel

--- a/optimum/neuron/distributed/base.py
+++ b/optimum/neuron/distributed/base.py
@@ -150,6 +150,58 @@ class Parallelizer(ABC):
             model, orig_to_parallel=orig_to_parallel, device=device, parallelize_embeddings=parallelize_embeddings
         )
         weight_map = getattr(model, "_weight_map", {})
+        # parameter2name = {id(p): n for n, p in model.named_parameters()}
+        # weights = defaultdict(lambda: {"modules": [], "concrete_parameter": None})
+        # to_visit = [model]
+        # modules_to_initialize = []
+        # visited = set()
+        # while to_visit:
+        #     mod = to_visit.pop(0)
+        #     if mod in visited:
+        #         continue
+        #     visited.add(mod)
+        #     for param in mod.parameters():
+        #         name = parameter2name[id(param)]
+        #         split = name.rsplit(".", maxsplit=1)
+        #         module = model.get_submodule(split[0])
+        #         attribute_name = split[1]
+        #         weights[id(param)]["modules"].append((module, attribute_name))
+        #         if weights[id(param)]["concrete_parameter"] is None:
+        #             if param.device == torch.device("meta"):
+        #                 if weight_map is None:
+        #                     raise ValueError(
+        #                         f"The parameter called {name} of the model is on the `meta` device and no weight map is "
+        #                         "attached to the model to load the proper weights from file."
+        #                     )
+        #             try:
+        #                 weight_info = WeightInformation(weight_map[name], name, device=device)
+        #                 concrete_parameter = torch.nn.Parameter(load_tensor_for_weight(weight_info))
+        #             except KeyError:
+        #                 # This means that there is no information about where to find the weights for this parameter.
+        #                 device = torch.device("cpu") if device is None else device
+        #                 concrete_parameter = torch.nn.Parameter(torch.empty_like(getattr(module, attribute_name), device=device))
+        #                 modules_to_initialize.append(module)
+        #             weights[id(param)]["concrete_parameter"] = concrete_parameter
+        #     to_visit.extend(mod.modules())
+
+        # for param, entry in weights.items():
+        #     print(entry["modules"])
+
+        # with torch.no_grad():
+        #     for parameter, entry in weights.items():
+        #         name = parameter2name[parameter]
+        #         associated_modules = entry["modules"]
+        #         concrete_parameter = entry["concrete_parameter"]
+        #         for module, attribute_name in associated_modules:
+        #             # This must be either a torch.nn.Embedding or a torch.nn.Linear since those are the only
+        #             # classes that we initialize on the `meta` device.
+        #             setattr(module, attribute_name, concrete_parameter)
+
+        #         for mod in modules_to_initialize:
+        #             # This module has not pre-trained weights, it must be fine-tuned, we initialize it with the
+        #             # `reset_parameters()` method.
+        #             mod.reset_parameters()
+
         with torch.no_grad():
             modules_to_initialize = []
             for name, parameter in model.named_parameters():
@@ -162,12 +214,8 @@ class Parallelizer(ABC):
                             "attached to the model to load the proper weights from file."
                         )
                     split = name.rsplit(".", maxsplit=1)
-                    if len(split) == 1:
-                        module = model
-                        attribute_name = split[0]
-                    else:
-                        module = model.get_submodule(split[0])
-                        attribute_name = split[1]
+                    module = model.get_submodule(split[0])
+                    attribute_name = split[1]
                     try:
                         weight_info = WeightInformation(weight_map[name], name, device=device)
                         setattr(module, attribute_name, torch.nn.Parameter(load_tensor_for_weight(weight_info)))

--- a/optimum/neuron/distributed/base.py
+++ b/optimum/neuron/distributed/base.py
@@ -150,58 +150,6 @@ class Parallelizer(ABC):
             model, orig_to_parallel=orig_to_parallel, device=device, parallelize_embeddings=parallelize_embeddings
         )
         weight_map = getattr(model, "_weight_map", {})
-        # parameter2name = {id(p): n for n, p in model.named_parameters()}
-        # weights = defaultdict(lambda: {"modules": [], "concrete_parameter": None})
-        # to_visit = [model]
-        # modules_to_initialize = []
-        # visited = set()
-        # while to_visit:
-        #     mod = to_visit.pop(0)
-        #     if mod in visited:
-        #         continue
-        #     visited.add(mod)
-        #     for param in mod.parameters():
-        #         name = parameter2name[id(param)]
-        #         split = name.rsplit(".", maxsplit=1)
-        #         module = model.get_submodule(split[0])
-        #         attribute_name = split[1]
-        #         weights[id(param)]["modules"].append((module, attribute_name))
-        #         if weights[id(param)]["concrete_parameter"] is None:
-        #             if param.device == torch.device("meta"):
-        #                 if weight_map is None:
-        #                     raise ValueError(
-        #                         f"The parameter called {name} of the model is on the `meta` device and no weight map is "
-        #                         "attached to the model to load the proper weights from file."
-        #                     )
-        #             try:
-        #                 weight_info = WeightInformation(weight_map[name], name, device=device)
-        #                 concrete_parameter = torch.nn.Parameter(load_tensor_for_weight(weight_info))
-        #             except KeyError:
-        #                 # This means that there is no information about where to find the weights for this parameter.
-        #                 device = torch.device("cpu") if device is None else device
-        #                 concrete_parameter = torch.nn.Parameter(torch.empty_like(getattr(module, attribute_name), device=device))
-        #                 modules_to_initialize.append(module)
-        #             weights[id(param)]["concrete_parameter"] = concrete_parameter
-        #     to_visit.extend(mod.modules())
-
-        # for param, entry in weights.items():
-        #     print(entry["modules"])
-
-        # with torch.no_grad():
-        #     for parameter, entry in weights.items():
-        #         name = parameter2name[parameter]
-        #         associated_modules = entry["modules"]
-        #         concrete_parameter = entry["concrete_parameter"]
-        #         for module, attribute_name in associated_modules:
-        #             # This must be either a torch.nn.Embedding or a torch.nn.Linear since those are the only
-        #             # classes that we initialize on the `meta` device.
-        #             setattr(module, attribute_name, concrete_parameter)
-
-        #         for mod in modules_to_initialize:
-        #             # This module has not pre-trained weights, it must be fine-tuned, we initialize it with the
-        #             # `reset_parameters()` method.
-        #             mod.reset_parameters()
-
         with torch.no_grad():
             modules_to_initialize = []
             for name, parameter in model.named_parameters():

--- a/optimum/neuron/distributed/base.py
+++ b/optimum/neuron/distributed/base.py
@@ -94,6 +94,7 @@ class Parallelizer(ABC):
         model: "PreTrainedModel",
         orig_to_parallel: Optional[Dict[int, "torch.nn.Parameter"]] = None,
         device: Optional["torch.device"] = None,
+        parallelize_embeddings: bool = True,
     ) -> "PreTrainedModel":
         """
         Parallelizes the model by transforming regular layer into their parallel counterparts.
@@ -107,6 +108,9 @@ class Parallelizer(ABC):
                 It might be deprecated soon.
             device (`Optional[torch.device]`, defaults to `None`):
                 The device where the new parallel layers should be put.
+            parallelize_embeddings (`bool`, defaults to `True`):
+                Whether or not the embeddings should be parallelized.
+                This can be disabled in the case when the TP size does not divide the vocabulary size.
 
         Returns:
             `PreTrainedModel`: The parallelized model.
@@ -118,6 +122,7 @@ class Parallelizer(ABC):
         model: "PreTrainedModel",
         orig_to_parallel: Optional[Dict[int, "torch.nn.Parameter"]] = None,
         device: Optional["torch.device"] = None,
+        parallelize_embeddings: bool = True,
     ) -> "PreTrainedModel":
         """
         Parallelizes the model by transforming regular layer into their parallel counterparts using
@@ -134,11 +139,16 @@ class Parallelizer(ABC):
                 It might be deprecated soon.
             device (`Optional[torch.device]`, defaults to `None`):
                 The device where the new parallel layers should be put.
+            parallelize_embeddings (`bool`, defaults to `True`):
+                Whether or not the embeddings should be parallelized.
+                This can be disabled in the case when the TP size does not divide the vocabulary size.
 
         Returns:
             `PreTrainedModel`: The parallelized model.
         """
-        model = cls._parallelize(model, orig_to_parallel=orig_to_parallel, device=device)
+        model = cls._parallelize(
+            model, orig_to_parallel=orig_to_parallel, device=device, parallelize_embeddings=parallelize_embeddings
+        )
         weight_map = getattr(model, "_weight_map", {})
         with torch.no_grad():
             modules_to_initialize = []

--- a/optimum/neuron/distributed/decoder_models.py
+++ b/optimum/neuron/distributed/decoder_models.py
@@ -51,8 +51,10 @@ class GPTNeoParallelizer(Parallelizer):
         model: "PreTrainedModel",
         orig_to_parallel: Optional[Dict[int, "torch.nn.Parameter"]],
         device: Optional["torch.device"] = None,
+        parallelize_embeddings: bool = True,
     ) -> "PreTrainedModel":
-        model = GPTNeoParallelEmbedding.transform(model, model, device=device)
+        if parallelize_embeddings:
+            model = GPTNeoParallelEmbedding.transform(model, model, device=device)
         for block in model.transformer.h:
             block.attn.attention = GPTNeoParallelSelfAttention.transform(
                 model,
@@ -130,8 +132,10 @@ class LlamaParallelizer(Parallelizer):
         model: "PreTrainedModel",
         orig_to_parallel: Optional[Dict[int, "torch.nn.Parameter"]],
         device: Optional["torch.device"] = None,
+        parallelize_embeddings: bool = True,
     ) -> "PreTrainedModel":
-        model = LlamaParallelEmbedding.transform(model, model, device=device)
+        if parallelize_embeddings:
+            model = LlamaParallelEmbedding.transform(model, model, device=device)
         for layer in model.model.layers:
             layer.self_attn = LlamaParallelSelfAttention.transform(model, layer.self_attn, device=device)
             layer.mlp = LLamaParallelMLP.transform(model, layer.mlp, device=device)

--- a/optimum/neuron/distributed/encoder_decoder_models.py
+++ b/optimum/neuron/distributed/encoder_decoder_models.py
@@ -133,8 +133,10 @@ class T5Parallelizer(Parallelizer):
         model: "PreTrainedModel",
         orig_to_parallel: Optional[Dict[int, "torch.nn.Parameter"]],
         device: Optional["torch.device"] = None,
+        parallelize_embeddings: bool = True,
     ) -> "PreTrainedModel":
-        model = T5ParallelEmbedding.transform(model, model, device=device)
+        if parallelize_embeddings:
+            model = T5ParallelEmbedding.transform(model, model, device=device)
         if model.encoder.embed_tokens is not None:
             model.encoder.embed_tokens = model.shared
         if model.decoder.embed_tokens is not None:

--- a/optimum/neuron/distributed/encoder_models.py
+++ b/optimum/neuron/distributed/encoder_models.py
@@ -72,6 +72,10 @@ class BertParallelizer(Parallelizer):
 class RobertaParallelEmbedding(ParallelEmbedding):
     EMBEDDING_NAME = "roberta.embeddings.word_embeddings"
     LM_HEAD_NAME = "lm_head.decoder"
+    LM_HEAD_NAME = {
+        "RobertaForCausalLM": "lm_head.decoder",
+        "RobertaForMaskedLM": "lm_head.decoder",
+    }
 
 
 class RobertaParallelSelfAttention(BertParallelSelfAttention):

--- a/optimum/neuron/distributed/encoder_models.py
+++ b/optimum/neuron/distributed/encoder_models.py
@@ -17,12 +17,17 @@
 from typing import TYPE_CHECKING, Dict, Optional
 
 from .base import Parallelizer
-from .parallel_layers import ParallelSelfAttention, ParallelSelfOutput
+from .parallel_layers import ParallelEmbedding, ParallelSelfAttention, ParallelSelfOutput
 
 
 if TYPE_CHECKING:
     import torch
     from transformers import PreTrainedModel
+
+
+class BertParallelEmbedding(ParallelEmbedding):
+    EMBEDDING_NAME = "bert.embeddings.word_embeddings"
+    LM_HEAD_NAME = "cls.predictions.decoder"
 
 
 class BertParallelSelfAttention(ParallelSelfAttention):
@@ -40,7 +45,11 @@ class BertParallelizer(Parallelizer):
         model: "PreTrainedModel",
         orig_to_parallel: Optional[Dict[int, "torch.nn.Parameter"]],
         device: Optional["torch.device"] = None,
+        parallelize_embeddings: bool = True,
     ) -> "PreTrainedModel":
+        # if parallelize_embeddings:
+        #     print("HEYYYYYYYYY")
+        #     model = BertParallelEmbedding.transform(model, model, device=device)
         for layer in model.bert.encoder.layer:
             layer.attention.self = BertParallelSelfAttention.transform(
                 model,
@@ -55,6 +64,11 @@ class BertParallelizer(Parallelizer):
                 device=device,
             )
         return model
+
+
+class RobertaParallelEmbedding(ParallelEmbedding):
+    EMBEDDING_NAME = "roberta.embeddings.word_embeddings"
+    LM_HEAD_NAME = "lm_head.decoder"
 
 
 class RobertaParallelSelfAttention(BertParallelSelfAttention):
@@ -72,7 +86,10 @@ class RobertaParallelizer(Parallelizer):
         model: "PreTrainedModel",
         orig_to_parallel: Optional[Dict[int, "torch.nn.Parameter"]],
         device: Optional["torch.device"] = None,
+        parallelize_embeddings: bool = True,
     ) -> "PreTrainedModel":
+        if parallelize_embeddings:
+            model = RobertaParallelEmbedding.transform(model, model, device=device)
         for layer in model.roberta.encoder.layer:
             layer.attention.self = RobertaParallelSelfAttention.transform(
                 model,

--- a/optimum/neuron/distributed/encoder_models.py
+++ b/optimum/neuron/distributed/encoder_models.py
@@ -27,7 +27,11 @@ if TYPE_CHECKING:
 
 class BertParallelEmbedding(ParallelEmbedding):
     EMBEDDING_NAME = "bert.embeddings.word_embeddings"
-    LM_HEAD_NAME = "cls.predictions.decoder"
+    LM_HEAD_NAME = {
+        "BertForPreTraining": "cls.predictions.decoder",
+        "BertLMHeadModel": "cls.predictions.decoder",
+        "BertForMaskedLM": "cls.predictions.decoder",
+    }
 
 
 class BertParallelSelfAttention(ParallelSelfAttention):
@@ -47,9 +51,8 @@ class BertParallelizer(Parallelizer):
         device: Optional["torch.device"] = None,
         parallelize_embeddings: bool = True,
     ) -> "PreTrainedModel":
-        # if parallelize_embeddings:
-        #     print("HEYYYYYYYYY")
-        #     model = BertParallelEmbedding.transform(model, model, device=device)
+        if parallelize_embeddings:
+            model = BertParallelEmbedding.transform(model, model, device=device)
         for layer in model.bert.encoder.layer:
             layer.attention.self = BertParallelSelfAttention.transform(
                 model,

--- a/optimum/neuron/distributed/encoder_models.py
+++ b/optimum/neuron/distributed/encoder_models.py
@@ -71,7 +71,6 @@ class BertParallelizer(Parallelizer):
 
 class RobertaParallelEmbedding(ParallelEmbedding):
     EMBEDDING_NAME = "roberta.embeddings.word_embeddings"
-    LM_HEAD_NAME = "lm_head.decoder"
     LM_HEAD_NAME = {
         "RobertaForCausalLM": "lm_head.decoder",
         "RobertaForMaskedLM": "lm_head.decoder",

--- a/optimum/neuron/distributed/parallel_layers.py
+++ b/optimum/neuron/distributed/parallel_layers.py
@@ -18,7 +18,7 @@ from abc import ABC, abstractclassmethod
 from pathlib import Path
 from typing import TYPE_CHECKING, Dict, Optional, Tuple, Union
 
-from ...utils import NormalizedConfigManager
+from ...utils import NormalizedConfigManager, logging
 from ..utils import is_neuronx_distributed_available
 from .utils import WeightInformation, embedding_to_parallel_embedding, linear_to_parallel_linear
 
@@ -29,6 +29,9 @@ if is_neuronx_distributed_available():
 if TYPE_CHECKING:
     import torch
     from transformers import PretrainedConfig, PreTrainedModel
+
+
+logger = logging.get_logger()
 
 
 class ParallelLayer(ABC):
@@ -167,6 +170,18 @@ class ParallelEmbedding(ParallelLayer):
                     lm_head_bias_weight_info = WeightInformation(
                         weight_map[lm_head_bias_weight_name], lm_head_bias_weight_name, device=device
                     )
+
+        embedding_layer = layer.get_submodule(cls.EMBEDDING_NAME)
+        tp_size = parallel_state.get_tensor_model_parallel_size()
+        if embedding_layer.num_embeddings % tp_size != 0:
+            import torch_xla.core.xla_model as xm
+
+            if xm.get_ordinal() == 0:
+                logger.warning(
+                    f"Embedding parallelization for TP was skipped because the tensor parallel size ({tp_size}) does not "
+                    f"divide the number of embeddings ({embedding_layer.num_embeddings})"
+                )
+            return layer
 
         parallel_layers = embedding_to_parallel_embedding(
             layer.get_submodule(cls.EMBEDDING_NAME),

--- a/optimum/neuron/distributed/parallel_layers.py
+++ b/optimum/neuron/distributed/parallel_layers.py
@@ -16,7 +16,7 @@
 
 from abc import ABC, abstractclassmethod
 from pathlib import Path
-from typing import TYPE_CHECKING, Dict, Optional, Tuple
+from typing import TYPE_CHECKING, Dict, Optional, Tuple, Union
 
 from ...utils import NormalizedConfigManager
 from ..utils import is_neuronx_distributed_available
@@ -106,12 +106,13 @@ class ParallelEmbedding(ParallelLayer):
     Attributes:
         EMBEDDING_NAME (`str`, defaults to `"embedding"`):
             The qualified name of the embedding layer.
-        LM_HEAD_NAME (`Optional[str]`, defaults to `None`):
-            The qualified name of the LM head tied to the embedding layer (if any).
+        LM_HEAD_NAME (`Optional[Union[str, Dict[str, str]]]`, defaults to `None`):
+            The qualified name of the LM head tied to the embedding layer (if any). It can be also a dictionary mapping
+            a class name to LM head qualified name.
     """
 
     EMBEDDING_NAME: str = "embedding"
-    LM_HEAD_NAME: Optional[str] = None
+    LM_HEAD_NAME: Optional[Union[str, Dict[str, str]]] = None
 
     @classmethod
     def transform(
@@ -122,10 +123,16 @@ class ParallelEmbedding(ParallelLayer):
         device: Optional["torch.device"] = None,
     ) -> "torch.nn.Module":
         if cls.LM_HEAD_NAME is not None:
-            parent_lm_head_module, parent_lm_head_attribute_name = cls._get_module_and_attribute_name(
-                layer, cls.LM_HEAD_NAME
-            )
-            model_has_lm_head = hasattr(parent_lm_head_module, parent_lm_head_attribute_name)
+            if isinstance(cls.LM_HEAD_NAME, dict):
+                lm_head_name = cls.LM_HEAD_NAME.get(model.__class__.__name__, None)
+            else:
+                lm_head_name = cls.LM_HEAD_NAME
+            model_has_lm_head = False
+            if lm_head_name is not None:
+                parent_lm_head_module, parent_lm_head_attribute_name = cls._get_module_and_attribute_name(
+                    layer, lm_head_name
+                )
+                model_has_lm_head = hasattr(parent_lm_head_module, parent_lm_head_attribute_name)
         else:
             model_has_lm_head = False
 
@@ -147,11 +154,11 @@ class ParallelEmbedding(ParallelLayer):
             )
             if model_has_lm_head:
                 if layer_qualified_name:
-                    lm_head_weight_name = f"{layer_qualified_name}.{cls.LM_HEAD_NAME}.weight"
-                    lm_head_bias_weight_name = f"{layer_qualified_name}.{cls.LM_HEAD_NAME}.bias"
+                    lm_head_weight_name = f"{layer_qualified_name}.{lm_head_name}.weight"
+                    lm_head_bias_weight_name = f"{layer_qualified_name}.{lm_head_name}.bias"
                 else:
-                    lm_head_weight_name = f"{cls.LM_HEAD_NAME}.weight"
-                    lm_head_bias_weight_name = f"{cls.LM_HEAD_NAME}.bias"
+                    lm_head_weight_name = f"{lm_head_name}.weight"
+                    lm_head_bias_weight_name = f"{lm_head_name}.bias"
                 if lm_head_weight_name in weight_map:
                     lm_head_weight_info = WeightInformation(
                         weight_map[lm_head_weight_name], lm_head_weight_name, device=device
@@ -163,7 +170,7 @@ class ParallelEmbedding(ParallelLayer):
 
         parallel_layers = embedding_to_parallel_embedding(
             layer.get_submodule(cls.EMBEDDING_NAME),
-            lm_head_layer=layer.get_submodule(cls.LM_HEAD_NAME) if model_has_lm_head else None,
+            lm_head_layer=layer.get_submodule(lm_head_name) if model_has_lm_head else None,
             embedding_weight_info=embedding_weight_info,
             lm_head_weight_info=lm_head_weight_info,
             lm_head_bias_weight_info=lm_head_bias_weight_info,

--- a/optimum/neuron/distributed/utils.py
+++ b/optimum/neuron/distributed/utils.py
@@ -280,13 +280,14 @@ def linear_to_parallel_linear(
     kwargs["device"] = device
 
     parallel_linear_layer = parallel_linear_class(linear_layer.in_features, linear_layer.out_features, **kwargs)
+
     tp_rank = get_tensor_model_parallel_rank()
     row_size, col_size = parallel_linear_layer.weight.shape
 
     with torch.no_grad():
         if axis == "row":
             if embedding_weight_to_tie is not None:
-                parallel_linear_layer.weight = embedding_weight_to_tie
+                parallel_linear_layer.weight.copy_(embedding_weight_to_tie)
             elif linear_layer_weight_info is not None:
                 weight_data = load_tensor_for_weight(
                     linear_layer_weight_info,
@@ -306,7 +307,7 @@ def linear_to_parallel_linear(
             if linear_layer.bias is not None:
                 if linear_layer_bias_weight_info is not None:
                     bias_weight_data = load_tensor_for_weight(linear_layer_bias_weight_info)
-                    parallel_linear_layer.bias.data = bias_weight_data
+                    parallel_linear_layer.bias.copy_(bias_weight_data)
                 else:
                     parallel_linear_layer.bias.copy_(linear_layer.bias)
 
@@ -314,7 +315,7 @@ def linear_to_parallel_linear(
                     orig_to_parallel[id(linear_layer.bias)] = parallel_linear_layer.bias
         else:
             if embedding_weight_to_tie is not None:
-                parallel_linear_layer.weight = embedding_weight_to_tie
+                parallel_linear_layer.weight.copy_(embedding_weight_to_tie)
             elif linear_layer_weight_info is not None:
                 weight_data = load_tensor_for_weight(
                     linear_layer_weight_info,
@@ -347,8 +348,7 @@ def linear_to_parallel_linear(
                         linear_layer_bias_weight_info,
                         tensor_slices=tensor_slices,
                     )
-                    parallel_linear_layer.bias.data = bias_weight_data
-
+                    parallel_linear_layer.bias.copy_(bias_weight_data)
                 else:
                     if gather_output:
                         parallel_linear_layer.bias.copy_(linear_layer.bias)

--- a/optimum/neuron/distributed/utils.py
+++ b/optimum/neuron/distributed/utils.py
@@ -182,7 +182,7 @@ def embedding_to_parallel_embedding(
                     None,
                 ),
             )
-            parallel_embedding_layer.weight.data = weight_data
+            parallel_embedding_layer.weight.copy_(weight_data)
         else:
             parallel_embedding_layer.weight.copy_(
                 embedding_layer.weight[tp_rank * row_size : (tp_rank + 1) * row_size, :]

--- a/optimum/neuron/distributed/utils.py
+++ b/optimum/neuron/distributed/utils.py
@@ -287,7 +287,7 @@ def linear_to_parallel_linear(
     with torch.no_grad():
         if axis == "row":
             if embedding_weight_to_tie is not None:
-                parallel_linear_layer.weight.copy_(embedding_weight_to_tie)
+                parallel_linear_layer.weight = embedding_weight_to_tie
             elif linear_layer_weight_info is not None:
                 weight_data = load_tensor_for_weight(
                     linear_layer_weight_info,
@@ -296,7 +296,7 @@ def linear_to_parallel_linear(
                         (tp_rank * col_size, (tp_rank + 1) * col_size),
                     ),
                 )
-                parallel_linear_layer.weight.data = weight_data
+                parallel_linear_layer.weight.copy_(weight_data)
             elif linear_layer.weight.device != torch.device("meta"):
                 parallel_linear_layer.weight.copy_(
                     linear_layer.weight[:, tp_rank * col_size : (tp_rank + 1) * col_size]
@@ -315,7 +315,7 @@ def linear_to_parallel_linear(
                     orig_to_parallel[id(linear_layer.bias)] = parallel_linear_layer.bias
         else:
             if embedding_weight_to_tie is not None:
-                parallel_linear_layer.weight.copy_(embedding_weight_to_tie)
+                parallel_linear_layer.weight = embedding_weight_to_tie
             elif linear_layer_weight_info is not None:
                 weight_data = load_tensor_for_weight(
                     linear_layer_weight_info,
@@ -324,7 +324,7 @@ def linear_to_parallel_linear(
                         None,
                     ),
                 )
-                parallel_linear_layer.weight.data = weight_data
+                parallel_linear_layer.weight.copy_(weight_data)
 
             elif linear_layer.weight.device != torch.device("meta"):
                 parallel_linear_layer.weight.copy_(

--- a/optimum/neuron/training_args.py
+++ b/optimum/neuron/training_args.py
@@ -56,13 +56,16 @@ class NeuronTrainingArgumentsMixin:
         default=1, metadata={"help": "The number of replicas the model will be sharded on."}
     )
     disable_embedding_parallelization: bool = field(
-        default=False,
+        default=True,
         metadata={"help": "Whether or not the embedding parallelization when doing TP should be disabled."},
     )
 
     def __post_init__(self):
         # Patches accelerate.utils.imports.is_tpu_available to match `is_torch_xla_available`
         patch_accelerate_is_tpu_available()
+
+        if not self.disable_embedding_parallelization:
+            raise NotImplementedError("Disabling the parallelization of the embeddings is not fully supported yet.")
 
         if self.fsdp != "":
             # Disabling FSDP until next release because it is still very experimental and not validated.

--- a/optimum/neuron/training_args.py
+++ b/optimum/neuron/training_args.py
@@ -57,7 +57,7 @@ class NeuronTrainingArgumentsMixin:
     )
     disable_embedding_parallelization: bool = field(
         default=False,
-        metadata={"help": "The number of replicas the model will be sharded on."},
+        metadata={"help": "Whether or not the embedding parallelization when doing TP should be disabled."},
     )
 
     def __post_init__(self):

--- a/optimum/neuron/training_args.py
+++ b/optimum/neuron/training_args.py
@@ -57,7 +57,7 @@ class NeuronTrainingArgumentsMixin:
     )
     disable_embedding_parallelization: bool = field(
         default=False,
-        metadata={"help": "The number of replicas the model will be sharded on.", "action": "store_true"},
+        metadata={"help": "The number of replicas the model will be sharded on."},
     )
 
     def __post_init__(self):

--- a/optimum/neuron/training_args.py
+++ b/optimum/neuron/training_args.py
@@ -55,6 +55,10 @@ class NeuronTrainingArgumentsMixin:
     tensor_parallel_size: int = field(
         default=1, metadata={"help": "The number of replicas the model will be sharded on."}
     )
+    disable_embedding_parallelization: bool = field(
+        default=False,
+        metadata={"help": "The number of replicas the model will be sharded on.", "action": "store_true"},
+    )
 
     def __post_init__(self):
         # Patches accelerate.utils.imports.is_tpu_available to match `is_torch_xla_available`
@@ -86,7 +90,7 @@ class NeuronTrainingArgumentsMixin:
                     "The minimal required Transformers version to perform XLA FSDP is "
                     f"{TRANSFORMERS_MIN_VERSION_FOR_XLA_FSDP} but {transformers.__version__} is installed."
                 )
-        self.tp_plugin = TensorParallelismPlugin(self.tensor_parallel_size)
+        self.tp_plugin = TensorParallelismPlugin(self.tensor_parallel_size, not self.disable_embedding_parallelization)
         super().__post_init__()
 
     # Needed only to specialize the warning message for FSDP.

--- a/tests/distributed/test_model_parallelization.py
+++ b/tests/distributed/test_model_parallelization.py
@@ -238,9 +238,10 @@ class ModelParallelizationTestCase(unittest.TestCase):
     def test_model_parallel_from_pretrained_without_lazy_load(self, model_class_name: str, model_name_or_path: str):
         self._test_model_parallel(model_class_name, model_name_or_path, False, False, True)
 
-    @parameterized.expand(MODELS_TO_TEST)
-    def test_model_parallel_without_parallelizing_embeddings(self, model_class_name: str, model_name_or_path: str):
-        self._test_model_parallel(model_class_name, model_name_or_path, False, False, False)
+    # TODO: enable that once it's working.
+    # @parameterized.expand(MODELS_TO_TEST)
+    # def test_model_parallel_without_parallelizing_embeddings(self, model_class_name: str, model_name_or_path: str):
+    #     self._test_model_parallel(model_class_name, model_name_or_path, False, False, False)
 
     # TODO: enable that.
     # @parameterized.expand(MODELS_TO_TEST)

--- a/tests/distributed/test_model_parallelization.py
+++ b/tests/distributed/test_model_parallelization.py
@@ -232,11 +232,15 @@ class ModelParallelizationTestCase(unittest.TestCase):
 
     @parameterized.expand(MODELS_TO_TEST)
     def test_model_parallel_from_config_without_lazy_load(self, model_class_name: str, model_name_or_path: str):
-        self._test_model_parallel(model_class_name, model_name_or_path, True, False, True)
+        self._test_model_parallel(
+            model_class_name, model_name_or_path, True, False, False
+        )  # Should be True once it's working.
 
     @parameterized.expand(MODELS_TO_TEST)
     def test_model_parallel_from_pretrained_without_lazy_load(self, model_class_name: str, model_name_or_path: str):
-        self._test_model_parallel(model_class_name, model_name_or_path, False, False, True)
+        self._test_model_parallel(
+            model_class_name, model_name_or_path, False, False, False
+        )  # Should be True once it's working.
 
     # TODO: enable that once it's working.
     # @parameterized.expand(MODELS_TO_TEST)

--- a/tests/distributed/test_model_parallelization.py
+++ b/tests/distributed/test_model_parallelization.py
@@ -112,6 +112,7 @@ class ModelParallelizationTestCase(unittest.TestCase):
         from_config: bool,
         tp_size: int,
         lazy_load: bool,
+        parallelize_embeddings: bool,
     ):
         model_import = f"from transformers import AutoConfig, AutoTokenizer, {model_class}"
         other_imports = (
@@ -157,7 +158,8 @@ class ModelParallelizationTestCase(unittest.TestCase):
         model_to_eval_mode = "model = model.eval()\nunsharded_model = unsharded_model.eval()"
 
         parallel_model_loading = (
-            "parallel_model = ParallelizersManager.parallelizer_for_model(model).parallelize(model)"
+            "parallel_model = ParallelizersManager.parallelizer_for_model(model).parallelize(model, "
+            f"parallelize_embeddings={parallelize_embeddings})"
         )
 
         inference = (
@@ -200,10 +202,15 @@ class ModelParallelizationTestCase(unittest.TestCase):
         )
 
     def _test_model_parallel(
-        self, model_class_name: str, model_name_or_path: str, from_config: bool, with_lazy_load: bool
+        self,
+        model_class_name: str,
+        model_name_or_path: str,
+        from_config: bool,
+        with_lazy_load: bool,
+        parallelize_embeddings: bool,
     ):
         python_code = self.get_parallel_test_python_file_content(
-            model_class_name, model_name_or_path, from_config, 2, with_lazy_load
+            model_class_name, model_name_or_path, from_config, 2, with_lazy_load, parallelize_embeddings
         )
 
         with TemporaryDirectory() as tmpdirname:
@@ -225,11 +232,15 @@ class ModelParallelizationTestCase(unittest.TestCase):
 
     @parameterized.expand(MODELS_TO_TEST)
     def test_model_parallel_from_config_without_lazy_load(self, model_class_name: str, model_name_or_path: str):
-        self._test_model_parallel(model_class_name, model_name_or_path, True, False)
+        self._test_model_parallel(model_class_name, model_name_or_path, True, False, True)
 
     @parameterized.expand(MODELS_TO_TEST)
     def test_model_parallel_from_pretrained_without_lazy_load(self, model_class_name: str, model_name_or_path: str):
-        self._test_model_parallel(model_class_name, model_name_or_path, False, False)
+        self._test_model_parallel(model_class_name, model_name_or_path, False, False, True)
+
+    @parameterized.expand(MODELS_TO_TEST)
+    def test_model_parallel_without_parallelizing_embeddings(self, model_class_name: str, model_name_or_path: str):
+        self._test_model_parallel(model_class_name, model_name_or_path, False, False, False)
 
     # TODO: enable that.
     # @parameterized.expand(MODELS_TO_TEST)

--- a/tests/distributed/test_utils.py
+++ b/tests/distributed/test_utils.py
@@ -202,8 +202,8 @@ class ParallelUtilsTestCase(unittest.TestCase):
             weight_filename = tmpdir / "weights.safetensors"
 
             if with_weight_info:
-                linear_weight = torch.randn(vocab_size, 300)
-                linear_bias_weight = torch.rand(vocab_size)
+                linear_weight = torch.randn(300, vocab_size) if axis == "row" else torch.randn(vocab_size, 300)
+                linear_bias_weight = torch.rand(300 if axis == "row" else vocab_size)
                 state_dict = {
                     "linear_weight": linear_weight,
                     "linear_bias_weight": linear_bias_weight,

